File: app/src/main/java/com/edufelip/finn/di/ViewModelModule.java
package com.edufelip.finn.di;

import com.edufelip.finn.data.network.ApiService;
import com.edufelip.finn.data.repositories.CommentRepository;
import com.edufelip.finn.data.repositories.CommunityRepository;
import com.edufelip.finn.domain.models.repositories.ICommentRepository;
import com.edufelip.finn.domain.models.repositories.ICommunityRepository;
import com.edufelip.finn.domain.models.repositories.IPostRepository;
import com.edufelip.finn.domain.models.repositories.IUserRepository;
import com.edufelip.finn.data.repositories.PostRepository;
import com.edufelip.finn.data.repositories.UserRepository;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.android.components.ViewModelComponent;
import dagger.hilt.android.scopes.ViewModelScoped;

@Module
@InstallIn(ViewModelComponent.class)
abstract class ViewModelModule {
    @Provides
    @ViewModelScoped
    static IUserRepository providesUserRepository(ApiService apiService) {
        return new UserRepository(apiService);
    }

    @Provides
    @ViewModelScoped
    static IPostRepository providesPostRepository(ApiService apiService) {
        return new PostRepository(apiService);
    }

    @Provides
    @ViewModelScoped
    static ICommunityRepository providesCommunityRepository(ApiService apiService) {
        return new CommunityRepository(apiService);
    }

    @Provides
    @ViewModelScoped
    static ICommentRepository providesCommentRepository(ApiService apiService) {
        return new CommentRepository(apiService);
    }
}

File: app/src/main/java/com/edufelip/finn/di/ViewModelModule.kt
package com.edufelip.finn.di

import com.edufelip.finn.shared.data.CacheTtlProvider
import com.edufelip.finn.shared.data.local.CommentCacheDataSource
import com.edufelip.finn.shared.data.local.CommunityCacheDataSource
import com.edufelip.finn.shared.data.local.PostCacheDataSource
import com.edufelip.finn.shared.data.remote.source.CommentRemoteDataSource
import com.edufelip.finn.shared.data.remote.source.CommunityRemoteDataSource
import com.edufelip.finn.shared.data.remote.source.PostRemoteDataSource
import com.edufelip.finn.shared.data.remote.source.UserRemoteDataSource
import com.edufelip.finn.shared.data.repository.DefaultCommentRepository
import com.edufelip.finn.shared.data.repository.DefaultCommunityRepository
import com.edufelip.finn.shared.data.repository.DefaultPostRepository
import com.edufelip.finn.shared.data.repository.DefaultUserRepository
import com.edufelip.finn.shared.domain.repository.CommentRepository
import com.edufelip.finn.shared.domain.repository.CommunityRepository
import com.edufelip.finn.shared.domain.repository.PostRepository
import com.edufelip.finn.shared.domain.repository.UserRepository
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.components.ViewModelComponent
import dagger.hilt.android.scopes.ViewModelScoped
import javax.inject.Named

@Module
@InstallIn(ViewModelComponent::class)
object ViewModelModule {
    @Provides
    @ViewModelScoped
    fun providesUserRepository(remote: UserRemoteDataSource): UserRepository =
        DefaultUserRepository(remote)

    @Provides
    @ViewModelScoped
    fun providesPostRepository(
        remote: PostRemoteDataSource,
        cacheDataSource: PostCacheDataSource,
        cacheTtlProvider: CacheTtlProvider,
    ): PostRepository = DefaultPostRepository(remote, cacheDataSource, cacheTtlProvider)

    @Provides
    @ViewModelScoped
    fun providesCommunityRepository(
        remote: CommunityRemoteDataSource,
        cacheDataSource: CommunityCacheDataSource,
        cacheTtlProvider: CacheTtlProvider,
    ): CommunityRepository = DefaultCommunityRepository(remote, cacheDataSource, cacheTtlProvider)

    @Provides
    @ViewModelScoped
    fun providesCommentRepository(
        remote: CommentRemoteDataSource,
        @Named("commentPageSize") pageSize: Int,
        cacheDataSource: CommentCacheDataSource,
        cacheTtlProvider: CacheTtlProvider,
    ): CommentRepository = DefaultCommentRepository(remote, cacheDataSource, cacheTtlProvider, pageSize)
}
